Experimental MenuButton

ValentinaKozlova opened the pull request:
### Platforms Impacted
- [ ] iOS
- [ ] macOS
- [X] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes
Moved MenuButton to the new framework.
It uses experimental Button component and Contextual menu.

Bug with icons will be fixed with ExperimentalButton redesign.

### Verification
![image](https://user-images.githubusercontent.com/11574680/131887553-bcc0faa0-ada1-4a0a-a12d-ec2e2439a1ab.png)

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [X] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

Saadnajmi commented:
High level question: 

While MenuButton is still new and we are onboarding people to use it, could we just not make a new package and just have the existing MenuButton use the new framework? I think we're at the point that we want new controls to use the new framework, and MenuButton is new enough that we could maybe just replace the old implementation before people start using it? For the macOS one, there is at least one client using MenuButton, but they don't customize it, so as long as the props are the same I think we could switch to the new framework without breaking API?

Saadnajmi commented on packages/experimental/MenuButton/src/renderContextualMenu.tsx:
```diff
@@ -0,0 +1,48 @@
+      return item;
+    });
+
+    return (<ContextualMenu {...contextualMenu}>
+          {menuItemsUpdated?.map((menuItem) => {
```
Indentation seems off. If you run `yarn prettier-fix` on this package do you still keep the indentation?

ValentinaKozlova replied:
I have it in my VSCode installed. I ran prettier-fix but it failed with one of the components

rurikoaraki replied:
Prettier should work now. I fixed it on Friday.

Saadnajmi commented on packages/experimental/MenuButton/src/renderContextualMenu.tsx:
```diff
@@ -0,0 +1,48 @@
+/* @jsxFrag React.Fragment */
+import React, { useState, useCallback } from 'react';
+import { ContextualMenu, ContextualMenuItem, SubmenuItem, Submenu, ContextualMenuProps } from '@fluentui-react-native/contextual-menu';
+import { MenuButtonItemProps } from './MenuButton.types'
+
+export const renderContextualMenu = (contextualMenu: ContextualMenuProps, menuItems: MenuButtonItemProps[]): JSX.Element => {
```
I'm guessing you made this a separate file so you could use the fragment? 
I like the separation, feels cleaner =)

Saadnajmi commented on packages/experimental/MenuButton/src/MenuButton.tsx:
```diff
@@ -0,0 +1,71 @@
+          <Button {...buttonProps}>
+            <Slots.chevronIcon />
+          </Button>
+          {showContextualMenu && renderContextualMenu(contextualMenuProps, menuItems)}
```
I just realized, why isn't the contextual menu just a slot?

ValentinaKozlova replied:
It didn't work with slots. The properties inside child components were undefined. To make it work I had to add property slotProps, but in this case I couldn't customize the props.

Saadnajmi replied:
undefined child components sounds like a bug we ran into when we forgot to add the JSX pragma "withSlots" at the top of the file. Could you elaborate on the bug you got trying to use this as a slot? I. suspect trying to use slots and fragments are conflicting, like you mentioned they were

ValentinaKozlova replied:
I didn't use fragments and slots in this case, because it even wasn't bundling. Some slots are working, for the icon for example, because I use there predefined slotProps

rurikoaraki replied:
I would prefer to make decisions on how to structure our components on whether it makes sense technically to do so, not whether it builds or bundles. We should be able to work through the building issues you mentioned if it makes sense to use slots.

Looking into Slots, it seems like the benefit it brings is that it filters out extra entries in the DOM created by child components that make up the larger component. This reduces overhead on the DOM when rendering.

It sounds like something that wouldn't be as useful for the Button but could save some entries in the context menu? Testing this somehow would be nice, since I'm mostly guessing here. I'll poke around with the render tree and see what results.

This is a pretty useful diagram (it's from the old framework, but the structure should be similar to the new one):
https://github.com/microsoft/fluentui-react-native/blob/4687ff096fd25b07e0a4e782616b6b2ec0fa2d9f/packages/deprecated/foundation-composable/docs/Composable.png

ValentinaKozlova replied:
The problem with bundling wasn't about using slots. We can't use in one file 'jsx' and 'use slots' pragmas. The promlem of slots in this case that customized properties were undefined.
What about extra entries in the DOM, I'm not sure. I checked the render tree for the ContextualMenu with slots and without, it looks the same.
Thank you for the diagram :)

rurikoaraki commented:
I tested the branch out and if I pass in a style which sets the menu button's background color to say, red, it doesn't seem to pick it up:

```
              <MenuButton
                primary
                icon={iconToShow}
                content="Primary MenuButton"
                menuItems={menuItems}
                onItemClick={onItemClick}
                contextualMenu={contextualMenuProps}
                style={{ backgroundColor: 'red' }}
              />
```

This seems like something that should work... what do you think?

rurikoaraki commented:
I'm also still trying to figure out what benefit/drawback there is for making or not making certain parts of the components Slots vs. not. Seems like with slots, we'd be able to utilize tokens better, but I haven't been able to fully form an understanding yet. Not a blocker, since if the current design works, then it's fine, but just something I'm thinking about.

Saadnajmi commented:
> I tested the branch out and if I pass in a style which sets the menu button's background color to say, red, it doesn't seem to pick it up:
> 
> ```
>               <MenuButton
>                 primary
>                 icon={iconToShow}
>                 content="Primary MenuButton"
>                 menuItems={menuItems}
>                 onItemClick={onItemClick}
>                 contextualMenu={contextualMenuProps}
>                 style={{ backgroundColor: 'red' }}
>               />
> ```
> 
> This seems like something that should work... what do you think?

I think this is one of those things were the API requires you use customize? Either that or we have to be better about merging inline styles that were passed in. I know something simple like this hasn't worked for other controls in the past too.

ValentinaKozlova commented:
> > I tested the branch out and if I pass in a style which sets the menu button's background color to say, red, it doesn't seem to pick it up:
> > ```
> >               <MenuButton
> >                 primary
> >                 icon={iconToShow}
> >                 content="Primary MenuButton"
> >                 menuItems={menuItems}
> >                 onItemClick={onItemClick}
> >                 contextualMenu={contextualMenuProps}
> >                 style={{ backgroundColor: 'red' }}
> >               />
> > ```
> > 
> > This seems like something that should work... what do you think?
> 
> I think this is one of those things were the API requires you use customize? Either that or we have to be better about merging inline styles that were passed in. I know something simple like this hasn't worked for other controls in the past too.

I think in this case it's not working because parent element for MenuButton is react.fragment. I'll try to pass style to the Button component and will check if it works

Saadnajmi commented on packages/experimental/MenuButton/src/MenuButton.tsx:
```diff
@@ -43,6 +43,7 @@ export const MenuButton = compose<MenuButtonType>({
       disabled,
       primary,
       icon,
+      style,
```
We don't need to do something fancy with merging the style generated by tokens with the user passed in style?

Saadnajmi replied:
This is a scenario where I wasn't sure what we're actually supposed to do.. so I'm not sure I can answer the question myself.

ValentinaKozlova commented:
> I tested the branch out and if I pass in a style which sets the menu button's background color to say, red, it doesn't seem to pick it up:
> 
> ```
>               <MenuButton
>                 primary
>                 icon={iconToShow}
>                 content="Primary MenuButton"
>                 menuItems={menuItems}
>                 onItemClick={onItemClick}
>                 contextualMenu={contextualMenuProps}
>                 style={{ backgroundColor: 'red' }}
>               />
> ```
> 
> This seems like something that should work... what do you think?

Yes, it should work. I fixed it, I added style prop to the Button component.

rurikoaraki commented on packages/experimental/MenuButton/src/__tests__/__snapshots__/MenuButton.test.win32.tsx.snap:
```diff
@@ -0,0 +1,55 @@
+  >
+    Press for Nested MenuButton
+  </Text>
+</View>
```
I'm curious why the chevron icon doesn't seem to show up in the tree

ValentinaKozlova replied:
Maybe because it's as SVG file here. I'll make it inline as in the old MenuButton

rurikoaraki commented:
Okay so after staring at the framework, I think I have a better understanding of how the slots work and why styling isn't applied if you use the slot method.

It seems that there's two "render passes" when something is rendered with compose. The first is building the components out of Slots using default props as defined by what's passed into compose for that component. The second seems to actually take the custom/user props into account (i.e. what is passed directly into the Slot in JSX). When we do the first pass, props are taken out of the slotProps provided into the compose function, and we will only run the "outer' portion of the render for each Slot. This is also the point where styling/tokens are taken into account. This means that, at the point where the Button is being created, the custom button props from the Menu button are not available, because those are only available in the second pass (when running the return method inside render). This is why things like the "primary" prop are missing when the button slot is built, and no styling takes effect. By the time the second render pass is going, styling from tokens have already been applied, so "primary" does nothing.

So basically because this current approach elects to re-use the Button's built-in tokens instead of providing new ones and feeding those in as tokens to the MenuButton, using Button as a non-slot is the only way we can get styling to be applied.

I think the main issue with this design, if I understand things correctly, is that the button tokens are not accessible by clients, so they can't try to change or override them. So for example if you want "primary" to have a different effect on the button for menu buttons, I don't know if the current design allows for that. I'm going to play with things to figure out the question to this, but wanted to post this explanation here in the meantime.

rurikoaraki commented:
> Okay so after staring at the framework, I think I have a better understanding of how the slots work and why styling isn't applied if you use the slot method.
> 
> It seems that there's two "render passes" when something is rendered with compose. The first is building the components out of Slots using default props as defined by what's passed into compose for that component. The second seems to actually take the custom/user props into account (i.e. what is passed directly into the Slot in JSX). When we do the first pass, props are taken out of the slotProps provided into the compose function, and we will only run the "outer' portion of the render for each Slot. This is also the point where styling/tokens are taken into account. This means that, at the point where the Button is being created, the custom button props from the Menu button are not available, because those are only available in the second pass (when running the return method inside render). This is why things like the "primary" prop are missing when the button slot is built, and no styling takes effect. By the time the second render pass is going, styling from tokens have already been applied, so "primary" does nothing.
> 
> So basically because this current approach elects to re-use the Button's built-in tokens instead of providing new ones and feeding those in as tokens to the Button from the MenuButton, using Button as a non-slot is the only way we can get styling to be applied.
> 
> I think the main issue with this design, if I understand things correctly, is that the button tokens are not accessible by clients when using MenyButton, so they can't try to change or override them. So for example if you want "primary" to have a different effect on the button for menu buttons, I don't know if the current design allows for that. I'm going to play with things to figure out the answer to this, but wanted to post this explanation here in the meantime.

So I tested out trying to customize the "primary" styling on the underlying button, and the current design doesn't allow for that. My guess would be that we need to expose tokens that we can apply to buttons as slot props. I would also suspect that you similarly can't customize the styling for other states like hover/pressed/disabled/etc.

However this is also the case with the menu button build with the older compose framework. At this point adding this doesn't remove any existing functionality and moves us closer to not using the older compose framwork for components, so I'd be okay with this getting committed since it's at parity with the older button, and having a new PR which addresses the issue above.

Saadnajmi commented:
I mentioned this offline, but I would remove the change that changes the svg from IconProps to SVGXML props. Then, we wouldn't be able to use other icon types for the icon.

ValentinaKozlova commented:
> I mentioned this offline, but I would remove the change that changes the svg from IconProps to SVGXML props. Then, we wouldn't be able to use other icon types for the icon.

I returned SVGProps

rurikoaraki commented on packages/experimental/MenuButton/src/MenuButton.types.ts:
```diff
@@ -0,0 +1,41 @@
+  context: MenuButtonContext;
+}
+
+export interface MenuButtonTokens extends FontTokens, IForegroundColorTokens, IBackgroundColorTokens, IBorderTokens {}
```
Been noticing this, but a suggestion that it may make styling/tokens easier to apply to the menu button in the future if MenuButtonTokens extended ButtonTokens, since it would make sense for a menu button to be as customizable as a button style-wise.
